packages/nextjs/components/scaffold-eth/Contract/ContractInput.tsx

```diff
@@ -0,0 +1,46 @@
+import { utils } from "ethers";
+import React, { Dispatch, SetStateAction, useMemo } from "react";
+
+import { AddressInput, Bytes32Input, BytesInput, InputBase, MissingTypeInput, Uint256Input } from "../Input";
+
+type ContractInputProps = {
+  setForm: Dispatch<SetStateAction<Record<string, any>>>;
+  form: Record<string, any>;
+  stateObjectKey: string;
+  paramType: utils.ParamType;
+};
+
+/**
+ * Generic Input component to handle input's based on their function param type
+ */
+export const ContractInput = ({ setForm, form, stateObjectKey, paramType }: ContractInputProps) => {
+  const inputProps = useMemo(
+    () => ({
+      name: stateObjectKey,
+      value: form[stateObjectKey],
+      placeholder: paramType.name ? `${paramType.type} ${paramType.name}` : paramType.type,
+      onChange: (value: any) => {
+        const formUpdate = { ...form };
+        formUpdate[stateObjectKey] = value;
+        setForm(formUpdate);
+      },
+    }),
+    [form, paramType.name, paramType.type, setForm, stateObjectKey],
```

Comment:
NITPICK: Since `form` here is an object, I think `useMemo` might be not usefull here since changes the dependencies are compared using [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) in react.

Reply:
But `form` is the output of a `useState` hook in the parent component, and we don't modify it except by using `setForm` so it's reference should be stable across renders, I believe.

Reply:
as I understand to use `useMemo` here, you need to memorize inputs on `ReadOnlyFunctionForm` and `WriteOnlyFunctionForm` since at least `setForm` will change on every render. And also form will change on every other input change too. Using global state for forms is pain, so you probably don't need useMemo here at all

Reply:
I removed it, looks like you guys were right. :+1: 
I generally manage more complex state with reducers or in the case of forms with `formik`.